Support complex numbers in eps

ChrisRackauckas opened the pull request:
This came up in https://github.com/JuliaMath/IterativeSolvers.jl/pull/113#issuecomment-301273365 . JuliaDiffEq and IterativeSolvers.jl have to make sure that the real-type is pulled out in order for `eps` to work:

```julia
eps(real(typeof(b)))
```

This detail can make many algorithms with tolerances that are written generically that would otherwise work with complex numbers error. This PR proposes to do just that trick, so that way `eps(1.0 + 1.0im)` returns machine epsilon for a Float64 (and generally works for `AbstractFloat` of course).

ChrisRackauckas committed:
Complex eps

musm commented:
btw I think these will need to be added to the complex.jl file since complex is not defined when float is loaded in sysimg

musm commented on base/float.jl:
```diff
@@ -713,6 +713,8 @@ end
     eps(::Type{Float32}) = $(bitcast(Float32, 0x34000000))
     eps(::Type{Float64}) = $(bitcast(Float64, 0x3cb0000000000000))
+    eps(::Type{Complex{T}}) where {T<:AbstractFloat} = eps(T)
+    eps(x::Complex{T}) where {T<:AbstractFloat} = eps(norm(x))
```
Is it the most appropriate to define the eps for  the norm of the number? E.g. why not  actually define it in terms of `eps(min(x.re,x.im)`  that way it gives the minimum next largest value?

musm replied:
in any case, whichever definition is chosen should be documented so it's clear

stevengj replied:
I would say to define `eps(z)` as follows:

* For any real number `x`, let `x̃ =fl(x)` be the closest floating-point value.  The key property of `ε=eps(x̃)` is that `|x - x̃| ≤ ε`.

* The corresponding thing should be true for complex numbers `z=a+ib` and the closest floating-point complex value `z̃=ã+ib̃=fl(a)+ifl(b)`:  we should have `|z - z̃| ≤ ε`.

This would suggest a definition `eps(z::Complex{T}) = hypot(eps(real(z)),eps(imag(z)))`.

stevengj replied:
But `eps(norm(z))` is a good approximation for my suggestion.

musm replied:
sounds reasonable,  @ChrisRackauckas  mind updating?
I guess this should also be added to Compat?

StefanKarpinski replied:
Why not go with the correct definition of `hypot(eps(real(z)), eps(imag(z)))` rather than the approximate value of `eps(norm(z))` and document this property as the defining characteristic for real and complex number types? IIRC, @simonbyrne already put some effort into documenting this very property for real values, so extending it to complexes makes sense.

StefanKarpinski replied:
Also, it's customary to use `z` for complex values and `x` for real values.

ChrisRackauckas committed:
Update float.jl

ChrisRackauckas committed:
Update complex.jl

ChrisRackauckas commented:
Made the change from `norm` to `hypot`. I added a few simple tests that spot check the value and the dispatches. I'll move the implementation to complex.jl if the tests fail. 

I'll be honest and say I put this PR in using the Github web interface and quickly put this in so I wouldn't forget. So this will probably need a squash in the end.

tkelman commented:
The docstring should probably also be updated to explain what the meaning is for complex inputs

musm commented:
bump (looks like this needs docstrings and to add the methods to the correct file for sys img loading)

musm commented on test/complex.jl:
```diff
@@ -871,6 +871,13 @@ end
 end
 
+@testset "eps" begin
+    @test eps(1.0+1.0im) == 3.1401849173675503e-16
```
might be better to use `===` to also test the types on the  lhs == rhs

ChrisRackauckas committed:
Move to complex.jl and add note to docstring

ChrisRackauckas committed:
remove empty doctest...

tkelman commented on base/float.jl:
```diff
@@ -737,6 +738,10 @@ julia> 1.0 + eps()
+
+For complex inputs `T`, `eps(T)` is defined as the distance bound to the nearest
+floating-point complex value, i.e. for z=a+ib and the closest floating-point
+complex value z̃=ã+ib̃=fl(a)+ifl(b), |z - z̃| ≤ ε/2.
```
should code or math highlight these

stevengj replied:
`fl(x)` is not a completely standard mathematical symbol (it is used in Trefethen's book and probably some other places), so we should probably state *where fl(x) is the closest floating-point value to a real x*.

We should probably also say |z - z̃| ≤ eps(z)/2.

ChrisRackauckas committed:
add math highlights

tkelman commented on base/float.jl:
```diff
@@ -740,8 +740,8 @@ julia> 1.0 + eps()/2
-floating-point complex value, i.e. for z=a+ib and the closest floating-point
-complex value z̃=ã+ib̃=fl(a)+ifl(b), |z - z̃| ≤ ε/2.
+floating-point complex value, i.e. for `z=a+ib` and the closest floating-point
+complex value `z̃=ã+ib̃=fl(a)+ifl(b)`, `|z - z̃| ≤ ε/2`.
```
math highlighting is double backticks

ChrisRackauckas committed:
change code highlight to math highlight

ChrisRackauckas committed:
make the eps(z) docstring more precise

musm commented on base/complex.jl:
```diff
@@ -897,6 +897,10 @@ function complex(A::AbstractArray{T}) where T
+
+## Machine epsilon for complex ##
+
+eps(z::Complex{T}) where {T<:AbstractFloat} = hypot(eps(real(z)), eps(imag(z)))
```
`hypot` is not defined till later in the system image load

So I think this should just be

```julia

function eps(z::Complex{T}) where {T<:AbstractFloat}
	x, y = eps(real(z)), eps(imag(z))
	if x < y
	   x, y = y, x
	end
	r = y/x
	x * sqrt(one(T) + r * r)
end

```

musm commented on base/float.jl:
```diff
@@ -713,6 +713,7 @@ end
     eps(::Type{Float16}) = $(bitcast(Float16, 0x1400))
     eps(::Type{Float32}) = $(bitcast(Float32, 0x34000000))
     eps(::Type{Float64}) = $(bitcast(Float64, 0x3cb0000000000000))
+    eps(::Type{Complex{T}}) where {T<:AbstractFloat} = eps(T)
```
this looks wrong

ChrisRackauckas replied:
yes that does

musm replied:
should probably just be
` sqrt(2*one(T))*eps(T)`

musm commented on test/complex.jl:
```diff
@@ -871,6 +871,13 @@ end
+@testset "eps" begin
+    @test eps(1.0+1.0im) == 3.1401849173675503e-16
+    @test eps(Complex128) == eps(1.0+1.0im)
+    @test eps(Complex64) == 1.1920929f-7
```
this will need to be updated too

ChrisRackauckas committed:
Use musm's inline hypot

ChrisRackauckas committed:
Update float.jl

ChrisRackauckas committed:
Update complex.jl

ChrisRackauckas commented:
Took the time to update this for @musm's comments.

stevengj commented on base/complex.jl:
```diff
@@ -897,6 +897,17 @@ function complex(A::AbstractArray{T}) where T
+
+function eps(z::Complex{T}) where {T<:AbstractFloat}
+    x, y = eps(real(z)), eps(imag(z))
+    if x < y
```
Maybe a comment that this is just `hypot(x,y)` and why you aren't calling `hypot`.

simonbyrne replied:
And documentation?

ChrisRackauckas replied:
This was already added to the eps docstring.

stevengj replied:
I don't understand why you have to inline `hypot` even if it is defined later.   Julia allows you to define functions "out of order" … all that matters is that `hypot` is defined by the time `eps(::Complex)` is actually called.

ChrisRackauckas replied:
@musm mentioned that's what the test failure was, so I gave it a try. I am not entirely sure what's going on with the tests since if I throw this into the REPL it works just fine.

ChrisRackauckas committed:
Update complex.jl

oscardssmith commented:
Bumping this due to https://discourse.julialang.org/t/eps-epsilon-for-complex-floats/59603/4. IMO for numbers, we should be using `max(eps(real(x)), eps(imag(x))`, but it might make sense to split this into `eps` for the type which is really easy, and for values which is less easy.

vtjnash commented:
I don't see anything for triage to discuss, it just needed tests

oscardssmith commented:
The thing I wanted triage input on is whether `hypot` is the behavior we want.

vtjnash commented on base/float.jl:
```diff
@@ -713,6 +713,7 @@ end
     eps(::Type{Float16}) = $(bitcast(Float16, 0x1400))
     eps(::Type{Float32}) = $(bitcast(Float32, 0x34000000))
     eps(::Type{Float64}) = $(bitcast(Float64, 0x3cb0000000000000))
+    eps(::Type{Complex{T}}) where {T<:AbstractFloat} = sqrt(2*one(T))*eps(T)
```
This should be in the complex.jl file also (or instead?–I'm not entirely sure which one was intended to be canonical)

vtjnash commented:
Triage won't decide something like that (source: I've been on the calls), as the relevant experts won't all be present haha. You should ping people here, where the discussion will reach the best people on their time!

JeffFessler commented:
I'm not a relevant expert, but I'll second @oscardssmith's suggestion of `max(eps(real(x)), eps(imag(x))` and note that matlab seems to be using `max`:
```matlab
>> [eps(1), eps(1 + 1i), eps(2), eps(1 + 2i), eps(2 + 1i)]

ans =
   1.0e-15 *
    0.2220    0.2220    0.4441    0.4441    0.4441
```

In contrast, as far as I can tell numpy considers only the type, not the value
https://github.com/numpy/numpy/blob/d44b9c61499f8bc5a9fc94286cd52f05e15e003f/numpy/lib/type_check.py#L387

```python
>>> from numpy.core import getlimits
>>> getlimits.finfo(1. + 100.j).eps
2.220446049250313e-16
>>> getlimits.finfo(0.).eps
2.220446049250313e-16
>>> numpy.finfo(110.).eps
2.220446049250313e-16
>>> numpy.finfo(100. + 1000.j).eps
2.220446049250313e-16
```

oscardssmith commented:
The big advantages I see `max` having over `hypot` is that it will always be an exact value, and that it is simpler to compute. I would be very interested if mathematical literature has anything to say on the subject.

StefanKarpinski commented:
For reals eps is how far you are from the next representable value. For complex numbers this would actually be the min of the real and imaginary eps values. It also has the familiar property that any point within eps/2 will round to the same value, which, as far as I can see is false for the hypot or max versions of this. What interpretation justifies the hypot definition? Smallest distance to a representable value that differs in both coordinates? Seems like a bizarre definition. What interpretation would justify the max version? I’m hard pressed to even think of one that makes sense.

vtjnash commented:
Looks like the original use case was for a tolerance, which I think make sense that it scales with the maximum or geometric average rounding error?

> Seems like a bizarre definition

Dunno, I had thought that was your suggestion (https://github.com/JuliaLang/julia/pull/21858#discussion_r117353723) haha

oscardssmith commented:
`hypot` gives the value such that all values within `eps/2` will round to the number. `max` gives the value such that all values within `eps/2` along either dimension will round to the number.

JeffFessler commented:
Using `min(abs(real(z)), abs(imag(z)))` would be problematic for numbers like `z = 0 + 1im`.

Keep in mind that `max` and `hypot` can differ at most by a factor of √2, and I'm hard pressed to think of a case where that small factor would make a difference.  I still prefer `max`, because it is always ≤ `hypot` (by at most √2) and therefore more "conservative" or "safer." 
To argue the other side, one could note that `isapprox` has a `norm` in it, which for a complex scalar is equivalent to the `hypot` being discussed here:
https://github.com/JuliaLang/julia/blob/f9720dc2ebd6cd9e3086365f281e62506444ef37/base/floatfuncs.jl#L278
Note that `isapprox` uses `sqrt(eps)` which is pretty typical of how `eps` is used and further illustrates the insignificance of a √2 factor either way.

JeffFessler commented:
@vtjnash following up on your earlier suggestion, who do we need to ping here to settle this?
Seems like the main question is `max` vs `hypot` (again I "vote" for `max`).
Once that is settled the rest of it (like tests and docstring) will be easy.
The initial PR on this topic was 4 years ago this week...  Happy birthday?

stevengj commented:
> I still prefer `max`, because it is always `≤ hypot` (by at most √2) and therefore more "conservative" or "safer."

Doesn't this make it *less* conservative?   `eps` usually represents an *upper* bound on the roundoff error (for rounding numbers to floating point).

oscardssmith commented:
This might be a stupid question, but can we just make it `eps(real(x))+im*eps(imag(x))`?

JeffFessler commented:
> Doesn't this make it _less_ conservative? `eps` usually represents an _upper_ bound on the |roundoff error|

I was thinking about it from the point of view of `isapprox`, where a smaller `eps` value will be a more stringent criterion for concluding that two complex numbers are practically equivalent.
But the `√2` factor is so unimportant one way or another that I'll gladly agree to `hypot` if it helps move this!
I agree that from a rounding perspective the `hypot` way is natural.


> can we just make it `eps(real(x)) + im*eps(imag(x))`

I could also live with the real-valued variant: `eps(real(x)) + eps(imag(x))` because that is a most a factor of 2 different from `max`.  But definitely `eps` should be real, not complex!

The docstring for `eps` in 1.6 says:

"Under the default `RoundNearest` rounding mode, if `y` is a real number and `x` is the nearest floating point number to `y`, then `|y-x| \leq \operatorname{eps}(x)/2.`"

To apply that same apply that same principle to complex numbers with `| |` interpreted as complex modulus, then it is easy to verify that `hypot` indeed is the appropriate method, as suggested by @StefanKarpinski 4 years ago.  So let's go with that and see if we can get this into 1.7?

stevengj commented:
I'm for `hypot` (I'm actually the one who suggested it in 2017 😉 ).

ChrisRackauckas commented:
Yeah `hypot` seems great. I don't know when I'll get to this but to me that seems like the way to go.

oscardssmith commented:
bump?

ChrisRackauckas commented:
I specialized my code for this in 2017 and am quite busy so I don't have any intention to finish this up "soon". If someone else wanted to complete this then please do. It should be rather quick but it's just a low enough priority that it probably won't happen for a bit.

ChrisRackauckas committed:
Use hypot in eps(::Complex) and move type method to complex.jl

Addresses review comments:
- Use hypot instead of inline implementation (JuliaLang/julia#21858 (comment))
- Move eps(::Type{Complex{T}}) from float.jl to complex.jl (JuliaLang/julia#21858 (comment))

stevengj commented on base/float.jl:
````diff
@@ -1064,6 +1064,11 @@ julia> 1.0 + eps()
 1.0
 ```
+
+For complex inputs `T`, `eps(T)` is defined as the distance bound to the nearest
+floating-point complex value, i.e. for ``z=a+ib`` and the closest floating-point
+complex value ``z̃=ã+ib̃=fl(a)+ifl(b)`` where ``fl(x)`` is the closest
+floating-point value to a real ``x``, ``eps(z)`` satisfies ``|z - z̃| ≤ eps(z)/2``.
````
It would be good to frame this as a strict generalization of the definition, since it is also true for real numbers, and then give complex numbers as a special case.

```suggestion
More generally, for any floating-point numeric type, `eps` corresponds to an
upper bound on the distance to the nearest floating-point complex value: if ``\text{fl}(x)`` is the closest
floating-point value to a number ``x`` (e.g. an arbitrary real number), then ``\text{fl}(x)``
satisfies ``|z - \text{fl}(x)| ≤ \text{eps}(z)/2``. This allows the definition of `eps` to be
extended to complex numbers, for which ``\text{fl}(a + ib) = \text{fl}(a) + i \text{fl}(b)``.
```

(Note that `eps(x)` and `fl(x)` should ideally be typeset with the AMS `\mathoperator` so that they aren't italicized, but that isn't supported in all LaTeX-math environments, and so I used `\text` instead to get $\text{eps}(x)$ and $\text{fl}(x)$.)

stevengj commented on base/complex.jl:
```diff
@@ -1141,3 +1141,16 @@ function complex(A::AbstractArray{T}) where T
+
+## Machine epsilon for complex ##
+
+function eps(z::Complex{T}) where {T<:AbstractFloat}
+    x, y = eps(real(z)), eps(imag(z))
+    hypot(x, y)
+end
```
Can be shortened to:
```suggestion
eps(z::Complex) = hypot(eps(real(z)), eps(imag(z)))
```

ChrisRackauckas committed:
Update base/complex.jl

Co-authored-by: Steven G. Johnson <stevenj@mit.edu>

ChrisRackauckas committed:
Update base/float.jl

Co-authored-by: Steven G. Johnson <stevenj@mit.edu>

ChrisRackauckas committed:
Merge branch 'master' into patch-4

giordano commented on test/complex.jl:
```diff
@@ -932,6 +932,13 @@ end
 
+@testset "eps" begin
+    @test eps(1.0+1.0im) === 3.1401849173675503e-16
+    @test eps(Complex128) === eps(1.0+1.0im)
+    @test eps(Complex64) === 1.6858739f-7
```
`Complex128` and `Complex64` don't exist (anymore)

JeffBezanson commented on base/complex.jl:
```diff
@@ -1141,3 +1141,13 @@ function complex(A::AbstractArray{T}) where T
+
+## Promotion to complex ##
+
+_default_type(T::Type{Complex}) = Complex{Int}
```
I think this should be removed? Doesn't seem to be used anywhere.
